### PR TITLE
PERF: RangeIndex.format performance

### PR DESCRIPTION
**Pull request by topper-123**
#35440 dropped ``RangeIndex._format_with_header``, which was functionally not needed, but needed to avoid creating an internal ndarray.

This rectifies that + gives some perf. improvements.

```python
>>> idx = pd.RangeIndex(1_000_000)
>>> %timeit idx.format()
4.6 s ± 102 ms per loop  # pandas v1.1.0
1.67 s ± 19.6 ms per loop  # master
595 ms ± 2.35 ms per loop  # this PR
```

Also, now the ``_data`` attribute isn't called, so this PR gives a perf. & memory improvement in some use cases compared to master:

```python
>>> idx = pd.RangeIndex(1_000_000)
>>> idx.format()
>>> "_data" in idx._cache
False # pandas v.1.1.0
True  # master
False  # this PR
```

### CODE REVIEW

**Review by arw2019 — COMMENTED** [review]
lgtm

Do we have benchmarks for this? If not do we want to add them?

File: pandas/core/indexes/range.py [hunk 1]
```diff
@@ -187,6 +187,11 @@ def _format_data(self, name=None):
         # we are formatting thru the attributes
         return None
 
+    def _format_with_header(self, header, na_rep="NaN") -> List[str]:
```
Comment (jreback) [inline pandas/core/indexes/range.py:190]: can you type

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
looks good. this is worth a perf mention in 1.2 whatsnew. ping on green.

**Comment by topper-123** [comment]
The tracisCI issue is just a timeout issue, but In `CI / Web and docs`, I've got a problem under  `Check ipython directive errors` that I'm not able to fix:

```
Run ! grep -B1 "^<<<-------------------------------------------------------------------------$" sphinx.log
[0 rows x 1 columns]
<<<-------------------------------------------------------------------------
##[error]Process completed with exit code 1.
```

The source code for this in `.github/workflows/ci.yml` says:

```yml
    # This can be removed when the ipython directive fails when there are errors,
    # including the `tee sphinx.log` in te previous step (https://github.com/ipython/ipython/issues/11547)
    - name: Check ipython directive errors
      run: "! grep -B1 \"^<<<-------------------------------------------------------------------------$\" sphinx.log"
```

I don't understand what this check is for. @TomAugspurger , you posted that mentioned ipython issue, do you mind looking at why my PR fails on this?

**Review by TomAugspurger — COMMENTED** [review]
There's an exception in the whatsnew/0.25.0.rst  file: https://github.com/pandas-dev/pandas/pull/35712/checks?check_run_id=1010393272#step:6:2305

File: pandas/core/indexes/range.py [hunk 2]
```diff
@@ -187,6 +187,15 @@ def _format_data(self, name=None):
         # we are formatting thru the attributes
         return None
 
+    def _format_with_header(self, header: List[str], na_rep: str = "NaN") -> List[str]:
+        if len(self._range) == 0:
+            return []
```
Comment (simonjayhawkins) [inline pandas/core/indexes/range.py:192]: should this be `return header`?

File: doc/source/whatsnew/v1.1.2.rst [hunk 3]
```diff
@@ -15,7 +15,7 @@ including other versions of pandas.
 Fixed regressions
 ~~~~~~~~~~~~~~~~~
 
--
+- Performance regression for :meth:`RangeIndex.format` (:issue:`35712`)
```
[the remark this answers is not in the document]
Comment (topper-123) [reply doc/source/whatsnew/v1.1.2.rst:18]: I looked briefly at it, wasn't sure how to do it, so I left it as is.

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/indexes/range.py]: Yes.

I added tests for this, and found that ``DateTimeIndex([]).format(name=True)`` returns ["None"], where it should return [""]. Same for PeriodIndex. I fixed that bug.

**Comment by simonjayhawkins** [comment]
@topper-123 merge conflicts in whatsnew

File: doc/source/whatsnew/v1.1.2.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.1.2.rst:18]: sure

**Review by simonjayhawkins — APPROVED** [review]
Thanks @topper-123 lgtm pending green

File: pandas/core/indexes/range.py [hunk 4]
```diff
@@ -187,6 +187,15 @@ def _format_data(self, name=None):
         # we are formatting thru the attributes
         return None
 
+    def _format_with_header(self, header: List[str], na_rep: str = "NaN") -> List[str]:
+        if len(self._range) == 0:
```
Comment (jreback) [inline pandas/core/indexes/range.py:191]: if not len(self._range)

**Review by jreback — APPROVED** [review]
minor comment, if you can address and merge on green.

**Comment by topper-123** [comment]
That merge brancg master has caused some strange error. anyone know how to fix it?

**Comment by topper-123** [comment]
Thanks, @simonjayhawkins, were there any git oneliners you used for this?

**Comment by simonjayhawkins** [comment]
> Thanks, @simonjayhawkins, were there any git oneliners you used for this?

I don't profess to know what was going on. resetting HEAD, merging master and then pulling from the branch seemed to fix except for the duplicate release note.

**Comment by simonjayhawkins** [comment]
> minor comment, if you can address and merge on green.

comment addressed

**Comment by simonjayhawkins** [comment]
Thanks @topper-123

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x

**Pull request merged**